File: apps/api/internal/perms/perms.go
// Sidcord permission bitmask sistemi (Discord modeline yakın)
package perms

// Permission bitleri — 64-bit bitmask
const (
	// Genel
	CreateInvite           uint64 = 1 << 0
	KickMembers            uint64 = 1 << 1
	BanMembers             uint64 = 1 << 2
	Administrator          uint64 = 1 << 3 // her şeyi yapabilir
	ManageChannels         uint64 = 1 << 4
	ManageGuild            uint64 = 1 << 5
	AddReactions           uint64 = 1 << 6
	ViewAuditLog           uint64 = 1 << 7
	PrioritySpeaker        uint64 = 1 << 8
	Stream                 uint64 = 1 << 9

	// Metin
	ViewChannel            uint64 = 1 << 10
	SendMessages           uint64 = 1 << 11
	SendTTSMessages        uint64 = 1 << 12
	ManageMessages         uint64 = 1 << 13
	EmbedLinks             uint64 = 1 << 14
	AttachFiles            uint64 = 1 << 15
	ReadMessageHistory     uint64 = 1 << 16
	MentionEveryone        uint64 = 1 << 17
	UseExternalEmojis      uint64 = 1 << 18
	ViewGuildInsights      uint64 = 1 << 19

	// Sesli
	Connect                uint64 = 1 << 20
	Speak                  uint64 = 1 << 21
	MuteMembers            uint64 = 1 << 22
	DeafenMembers          uint64 = 1 << 23
	MoveMembers            uint64 = 1 << 24
	UseVAD                 uint64 = 1 << 25 // voice activity detection

	// Diğer
	ChangeNickname         uint64 = 1 << 26
	ManageNicknames        uint64 = 1 << 27
	ManageRoles            uint64 = 1 << 28
	ManageWebhooks         uint64 = 1 << 29
	ManageEmojis           uint64 = 1 << 30
	UseApplicationCommands uint64 = 1 << 31
	RequestToSpeak         uint64 = 1 << 32
	ManageThreads          uint64 = 1 << 33
	CreatePublicThreads    uint64 = 1 << 34
	CreatePrivateThreads   uint64 = 1 << 35
	UseExternalStickers    uint64 = 1 << 36
	SendMessagesInThreads  uint64 = 1 << 37
	UseEmbeddedActivities  uint64 = 1 << 38
	ModerateMembers        uint64 = 1 << 39 // timeout
)

// Has — verilen permission'ı içeriyor mu (Administrator her şeyi geçer)
func Has(have, want uint64) bool {
	if have&Administrator != 0 {
		return true
	}
	return have&want == want
}

// Combine — birden fazla rolün bitmask'lerini birleştirir
func Combine(masks ...uint64) uint64 {
	var out uint64
	for _, m := range masks {
		out |= m
	}
	return out
}

// Tüm permission'lar (admin) — ayar kolaylığı için
const All = 0xFFFFFFFFFF

// Names — debug + UI için
var Names = map[uint64]string{
	CreateInvite:           "DAVET_OLUŞTUR",
	KickMembers:            "ÜYE_AT",
	BanMembers:             "ÜYE_BANLA",
	Administrator:          "YÖNETİCİ",
	ManageChannels:         "KANALLARI_YÖNET",
	ManageGuild:            "SUNUCUYU_YÖNET",
	AddReactions:           "TEPKİ_EKLE",
	ViewAuditLog:           "DENETİM_KAYDI_GÖR",
	PrioritySpeaker:        "ÖNCELİKLİ_KONUŞMACI",
	Stream:                 "YAYIN_AÇ",
	ViewChannel:            "KANAL_GÖR",
	SendMessages:           "MESAJ_GÖNDER",
	ManageMessages:         "MESAJ_YÖNET",
	EmbedLinks:             "BAĞLANTI_GÖMME",
	AttachFiles:            "DOSYA_EKLE",
	ReadMessageHistory:     "MESAJ_GEÇMİŞİ_OKU",
	MentionEveryone:        "EVERYONE_BAHSET",
	Connect:                "SESLİ_KATIL",
	Speak:                  "KONUŞ",
	MuteMembers:            "ÜYELERİ_SUSTUR",
	DeafenMembers:          "ÜYELERİ_SAĞIRLAŞTIR",
	MoveMembers:            "ÜYELERİ_TAŞI",
	UseVAD:                 "SES_AKTİVASYONU",
	ChangeNickname:         "TAKMA_AD_DEĞİŞTİR",
	ManageNicknames:        "TAKMA_ADLARI_YÖNET",
	ManageRoles:            "ROLLERİ_YÖNET",
	ManageWebhooks:         "WEBHOOK'LARI_YÖNET",
	ManageEmojis:           "EMOJİLERİ_YÖNET",
	UseApplicationCommands: "SLASH_KOMUTU_KULLAN",
	ModerateMembers:        "MODERATION_TIMEOUT",
}

File: apps/web/src/perms.ts
// Frontend permission bitmask sabitleri (backend ile aynı)
// 64-bit bitmask — JS Number 53-bit ile sınırlı, BigInt kullanıyoruz
export const PERM = {
  CREATE_INVITE: 1n << 0n,
  KICK_MEMBERS: 1n << 1n,
  BAN_MEMBERS: 1n << 2n,
  ADMINISTRATOR: 1n << 3n,
  MANAGE_CHANNELS: 1n << 4n,
  MANAGE_GUILD: 1n << 5n,
  ADD_REACTIONS: 1n << 6n,
  VIEW_AUDIT_LOG: 1n << 7n,
  PRIORITY_SPEAKER: 1n << 8n,
  STREAM: 1n << 9n,
  VIEW_CHANNEL: 1n << 10n,
  SEND_MESSAGES: 1n << 11n,
  SEND_TTS: 1n << 12n,
  MANAGE_MESSAGES: 1n << 13n,
  EMBED_LINKS: 1n << 14n,
  ATTACH_FILES: 1n << 15n,
  READ_HISTORY: 1n << 16n,
  MENTION_EVERYONE: 1n << 17n,
  USE_EXTERNAL_EMOJIS: 1n << 18n,
  VIEW_INSIGHTS: 1n << 19n,
  CONNECT: 1n << 20n,
  SPEAK: 1n << 21n,
  MUTE_MEMBERS: 1n << 22n,
  DEAFEN_MEMBERS: 1n << 23n,
  MOVE_MEMBERS: 1n << 24n,
  USE_VAD: 1n << 25n,
  CHANGE_NICKNAME: 1n << 26n,
  MANAGE_NICKNAMES: 1n << 27n,
  MANAGE_ROLES: 1n << 28n,
  MANAGE_WEBHOOKS: 1n << 29n,
  MANAGE_EMOJIS: 1n << 30n,
  USE_APP_COMMANDS: 1n << 31n,
  MODERATE_MEMBERS: 1n << 39n,
} as const;

export const PERM_LABELS: Record<string, string> = {
  CREATE_INVITE: 'Davet Oluştur',
  KICK_MEMBERS: 'Üye At',
  BAN_MEMBERS: 'Üye Banla',
  ADMINISTRATOR: 'Yönetici (her şey)',
  MANAGE_CHANNELS: 'Kanalları Yönet',
  MANAGE_GUILD: 'Sunucuyu Yönet',
  ADD_REACTIONS: 'Tepki Ekle',
  VIEW_AUDIT_LOG: 'Denetim Kaydı',
  VIEW_CHANNEL: 'Kanal Gör',
  SEND_MESSAGES: 'Mesaj Gönder',
  MANAGE_MESSAGES: 'Mesajları Yönet',
  EMBED_LINKS: 'Bağlantı Gömme',
  ATTACH_FILES: 'Dosya Ekle',
  READ_HISTORY: 'Mesaj Geçmişi Oku',
  MENTION_EVERYONE: '@everyone Bahset',
  CONNECT: 'Sesli Kanala Katıl',
  SPEAK: 'Konuş',
  MUTE_MEMBERS: 'Üyeleri Sustur',
  DEAFEN_MEMBERS: 'Üyeleri Sağırlaştır',
  MOVE_MEMBERS: 'Üyeleri Taşı',
  USE_VAD: 'Ses Aktivasyonu',
  CHANGE_NICKNAME: 'Takma Ad Değiştir',
  MANAGE_NICKNAMES: 'Takma Adları Yönet',
  MANAGE_ROLES: 'Rolleri Yönet',
  MANAGE_WEBHOOKS: 'Webhook Yönet',
  MANAGE_EMOJIS: 'Emojileri Yönet',
  MODERATE_MEMBERS: 'Timeout Uygula',
};

export function has(haveStr: string, want: bigint): boolean {
  const have = BigInt(haveStr || '0');
  if ((have & PERM.ADMINISTRATOR) !== 0n) return true;
  return (have & want) === want;
}

export function toggle(haveStr: string, perm: bigint): string {
  const have = BigInt(haveStr || '0');
  const next = (have & perm) !== 0n ? have & ~perm : have | perm;
  return next.toString();
}
